domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +192,305 @@ class TopicListController @Inject constructor(
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.STARTED_NOT_COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
-
+        val startedChapterProgressList = getStartedChapterProgressList(storyProgress)
         val recentlyPlayerChapterProgress: ChapterProgress? =
           startedChapterProgressList.firstOrNull()
-        if (recentlyPlayerChapterProgress != null) {
-          val recentlyPlayerChapterSummary: ChapterSummary? =
-            story.chapterList.find { chapterSummary ->
-              recentlyPlayerChapterProgress.explorationId == chapterSummary.explorationId
-            }
-          if (recentlyPlayerChapterSummary != null) {
-            val numberOfDaysPassed =
-              (Date().time - recentlyPlayerChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
-            val promotedStory = createPromotedStory(
+
+        checkIfStoryIsCompleted(topic.topicId, mostRecentCompletedChapterProgress, story)
+
+        when {
+          recentlyPlayerChapterProgress != null -> {
+            createOngoingStoryListBasedOnRecentlyPlayed(
               storyId,
+              story,
+              recentlyPlayerChapterProgress,
+              completedChapterProgressList,
               topic,
-              completedChapterProgressList.size,
-              story.chapterCount,
-              recentlyPlayerChapterSummary.name,
-              recentlyPlayerChapterSummary.explorationId
-            )
-            if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
-              ongoingStoryListBuilder.addRecentStory(promotedStory)
-            } else {
-              ongoingStoryListBuilder.addOlderStory(promotedStory)
+              completedStoryTopicId
+            ).let {
+              Pair(it.first, it.second)
+              numberOfDaysPassed = it.second
+              it.first?.let { it1 -> recentlyPlayedPromotedStoryList.add(it1) }
             }
           }
-        } else if (lastCompletedChapterProgress != null &&
-          lastCompletedChapterProgress.explorationId != story.chapterList.last().explorationId
-        ) {
-          val lastChapterSummary: ChapterSummary? = story.chapterList.find { chapterSummary ->
-            lastCompletedChapterProgress.explorationId == chapterSummary.explorationId
-          }
-          val nextChapterIndex = story.chapterList.indexOf(lastChapterSummary) + 1
-          val nextChapterSummary: ChapterSummary? = story.chapterList[nextChapterIndex]
-          if (nextChapterSummary != null) {
-            val numberOfDaysPassed =
-              (Date().time - lastCompletedChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
-            val promotedStory = createPromotedStory(
+          mostRecentCompletedChapterProgress != null &&
+            mostRecentCompletedChapterProgress.explorationId !=
+            story.chapterList.last().explorationId -> {
```

Comment:
Please add a brief comment explaining this logic (e.g. that we only compute the ongoing story list for stories that are fully completed yet).

Reply:
Done.